### ISSUE TITLE
tf.data.Dataset .map().batch() pattern is not matched to use fused implementation.

### ISSUE THREAD

**Issue by mcourteaux** [tensorflow/tensorflow#53572]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): v1.12.1-69264-g0cdf35562dc 2.9.0
- Python version: 3.8.10
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 11.5 / 8.3
- GPU model and memory: GTX1660 Ti

**Describe the current behavior**
combining `tf.data.Dataset.map()` with `.batch()` does not use the fused BatchAndMap implementation.

**Describe the expected behavior**
It does use the fused implementation. Currently, it's only possible to use the fused implementation when using the deprecated `experimental.map_and_batch()` transformation.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**

```py
import os
import datetime
from tqdm import tqdm
import numpy as np

import tensorflow as tf
print('TF version', tf.__version__)

gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    try:
        # Currently, memory growth needs to be the same across GPUs
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
        logical_gpus = tf.config.experimental.list_logical_devices('GPU')
        print(len(gpus), 'Physical GPUs,', len(logical_gpus), 'Logical GPUs')
    except RuntimeError as e:
        # Memory growth must be set before GPUs have been initialized
        print(e)


@tf.function
def do_stuff(wmat, tf_var):
    with tf.device("/gpu:0"):
        S = tf.constant(0.0)
        for i in tf.range(4):
            fi = tf.cast(i, dtype=tf.float32)
            A = tf.math.lgamma(tf.tanh(tf.matmul(wmat + fi, tf.transpose(wmat - fi, [0, 2, 1]))))
            S += tf.reduce_sum(A)
        error = tf.reduce_mean(tf_var)
        return error, S

exp_uuid = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")

n_batches = 512


def gen():
    for i in range(n_batches):
        with tf.device("/cpu:0"): # Make sure it comes from CPU
            r = tf.ones((400,800))
        yield r

option_names = ['map().batch()', 'map_and_batch()']
for option in range(2):

    with tf.device("/cpu:0"):
        dataset = tf.data.Dataset.from_generator(gen, output_types=tf.float32)

        def my_identity(x):
            with tf.device("/cpu:0"):
                print("my_identity input:", x, x.device)
                y = tf.identity(x)
                print("my_identity output:", y, y.device)
                return y

        if option == 0:
            ## Option 0: map().batch()
            dataset = dataset.map(my_identity).batch(16)

        elif option == 1:
            ## Option 1: deprecated map_and_batch()
            dataset = dataset.apply(tf.data.experimental.map_and_batch(my_identity, 16))

    gpu_transform = tf.data.experimental.prefetch_to_device('/gpu:0', buffer_size=4)
    dataset = dataset.apply(gpu_transform)


    tf_var = tf.Variable(tf.zeros(3))
    adam = tf.keras.optimizers.Adam(1e-4)
    logpath = os.path.join('data', 'logs', 'pa_' + exp_uuid + '_' + option_names[option])

    tf.profiler.experimental.start(logpath)
    start = datetime.datetime.now()
    for b, wmat in tqdm(enumerate(dataset)):
        with tf.GradientTape() as tape:

            if b == 0:
                print('\n\n dataset element device', wmat.device)
                print('\n')

            # Do some calculations
            result = do_stuff(wmat, tf_var)

        grads = tape.gradient(result[0], [tf_var])
        adam.apply_gradients(zip(grads, [tf_var]))
    stop = datetime.datetime.now()
    tf.profiler.experimental.stop()

    print(f'\n\nOption {option_names[option]}\n===========================\n')
    print(logpath)
    print('Time lapsed=', stop - start)
    print("\n\n")
```

**Other info / logs**

**Option 1**:
![image](https://user-images.githubusercontent.com/845012/147661299-a7f72017-00ff-47b6-bb71-8812bd5163d3.png)
Symptoms:
 - See the blocks `Iterator::FlapMap` and `Iterator::BatchV2` stacked on top of each other.
 - The MemcpyH2D (selected, see the details panel) is comping from pagable memory, instead of pinned memory (which is what MapAndBatch does). Because of the source being pagable memory, it can't overlap with kernel computations.
 
**Option 2**:
![image](https://user-images.githubusercontent.com/845012/147661558-9f316201-a4e5-4df1-a77b-032272537321.png)
Evidence:
 - The MapAndBatch block is used.
 - The MemcopyH2D comes from pinned memory (see details pane) and overlaps with kernel computations.

The whole deal about pinned memory is to allow parallel data upload and kernel computations. So the dataset needs to be produced into pinned host memory, which then can be uploaded asynchronously by the driver without an extra copy. See https://github.com/tensorflow/tensorflow/issues/43905#issuecomment-823675760 and https://github.com/tensorflow/tensorflow/issues/43905#issuecomment-824145184 and:
https://github.com/tensorflow/tensorflow/blob/40e9b534962989af7486bc6567ca472d71eb5049/tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op.cc#L522

This is a follow up on https://github.com/tensorflow/tensorflow/issues/43905.

**Comment by tilakrayal** [comment]
@mcourteaux ,
We see that you are using tf version 1.12, 1.x is not actively supported, please update to latest stable tf v2.7 and let us know if you are using same issue.

**Comment by mcourteaux** [comment]
No, I'm most definitely not. This was a fresh build from master branch from yesterday. Idk why the script that gives the TF version gives 1.12. It's most definitely wrong. I moved to TF 2 years ago.

**Comment by tilakrayal** [comment]
@mcourteaux ,
 I was able to execute the code without any issues.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/d0b954e177f03ab8e85b6fa087612ba6/untitled171.ipynb).Please provide the error log and also confirm if anything is missing here.Thanks!

**Comment by mcourteaux** [comment]
@tilakrayal I get the sensation you are not paying attention. How does Google hope to get contributions to a project if all of the useful feedback is dismissed as being either wrong or nobody paying attention? You somehow managed to make a notebook with the code from the linked issue, not mine. There is no error message, with my code. I showed you how the `MapAndBatchDataset` implementation is only used when using `tf.data.experimental.map_and_batch()`, and not with `map().batch()` as is actually promised. To see this, you look at the performance trace in TensorBoard, not in a log or error message.

**Comment by jsimsa** [comment]
Hi @mcourteaux, thank you for the detailed repro and sorry for the initial response. I will have someone on the tf.data team take a closer look.

**Comment by jsimsa** [comment]
@wilsingosti is taking a look

**Comment by wilsingosti** [comment]
The `my_identity` function is a no op. When `option = 1`, the `Map` is eliminated by the tf.data `noop_elimination` optimization. What is left is just `Batch`. If you replace the `my_identify` function with another that is not a no op, you should see the `MapAndBatch` implementation.

**Comment by jsimsa** [comment]
IIUC, the `map` in the repro above is introduced to make sure the memory allocated for the batch is [pinned](https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/). From that perspective, the `map` is intended to be a no-op. To keep the `map` is and still trigger the fusion, it should be sufficient to disable the no-op elimination using tf.data [options](https://www.tensorflow.org/api_docs/python/tf/data/Options).

**Comment by mcourteaux** [comment]
@wilsingosti Thanks for checking this. I'm wondering why not most Dataset implementations use the `gpu_compatible` memory when allocating stuff. Wouldn't that be useful in general, and then have us not rely on this `MapAndBatch` being selected to just copy the data to pinned memory.

**Comment by wilsingosti** [comment]
Yes, it would be useful in general. AFAIK, it has just not been prioritized so far. I will try to do this for `Batch` dataset.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by mcourteaux** [comment]
Don't close my issues. Go check if the operations by now used pinned memory or not. If they do, close it with a comment and commit that fixed it. If not, leave it open. Simply closing issues is not fixing anything. @wilsingosti was gonna take a look at fixing it. They might now if that's done or not by now.

**Comment by tripathi-genius** [comment]
Is the issue open? I want to contribute.

**Comment by Venkat6871** [comment]
Hi @mcourteaux ,
Thank you so much for filing this bug. We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!
